File: 0213.HouseRobberII.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n <= 1:
            return sum(nums)
        return max(self.dp(nums, 0, n-1), self.dp(nums, 1, n))

    def dp(self, nums, l, r):
        last, now = 0, 0
        for num in nums[l:r]:
            last, now = now, max(now, last+num)
        return now

File: 0000-0999/0213.HouseRobberII.go
func rob(nums []int) int {
    n := len(nums)
    if n == 0 {
        return 0
    } else if n == 1 {
        return nums[0]
    }
    return max(dp(nums, 0, n-1), dp(nums, 1, n))
}

func dp(nums []int, left, right int) int {
    pre, cur := 0, 0
    for i := left; i < right; i++ {
        pre, cur = cur, max(pre+nums[i], cur)
    }
    return  cur
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}